DDD-12.md

```diff
@@ -0,0 +1,349 @@
+```
+
+![](./DDD-12/s2-dbz-lib.png)
+
+## Quarkus Debezium Extension configuration
+
+The extensions must be configurable using the properties and yaml like any Quarkus application. The configuration properties available for the debezium engine must be available using a prefix `quarkus.debezium.xxx` like:
+
+```properties
+quarkus.debezium.configuration.offset.storage=org.apache.kafka.connect.storage.MemoryOffsetBackingStore
+quarkus.debezium.configuration.database.hostname=localhost
+quarkus.debezium.configuration.database.port=5432
+quarkus.debezium.configuration.database.user=postgresuser
+quarkus.debezium.configuration.database.password=postgrespw
+quarkus.debezium.configuration.database.dbname=postgresuser
+quarkus.debezium.configuration.snapshot.mode=never
+```
+
+Apart from the usual configuration properties like data source addresses, Debezium traditionally follows a _configuration over code_ approach, defining certain behavioral aspects using external configuration files interpreted at runtime. However, this approach changes in the Quarkus extension, which favors _code over configuration_—or more specifically, _annotation over configuration_. In this model, some features of the Debezium Engine are exposed through annotations, making Debezium instrumentation more expressive and developer-friendly.
+
+## Quarkus Debezium Extension DI
+
+Debezium internally use the `ServiceRegistry` to inject and manage object lifecycle thanks to `ServiceLoader` mechanism. Quarkus includes a lightweight CDI implementation called `ArC` which can be used to manage the classes loaded through the `ServiceLoader`.
+
+## Quarkus Debezium Extension additional feature
+
+The extension permits to address some use-cases already present in Debezium but in a _Quarkus_ way:
+
+- [1.Debezium Engine Instance](#quarkus-debezium-engine-instance)
+- [2.Debezium Engine Lifecycle events](#quarkus-debezium-lifecycle-events)
+- [3.Debezium Heartbeat events](#quarkus-debezium-heartbeats-events)
+- [4.Debezium Listener](#quarkus-debezium-listener)
+- [5.Debezium Custom Data Converter](#custom-debezium-data-converter)
+- [6.Debezium SchemaChange Listener](#quarkus-debezium-schemachange-listener)
+- [7.Debezium Notification Handler](#quarkus-debezium-notification-handler)
+- [8.Debezium PostProcess Consumer](#quarkus-debezium-postprocessor)
+- [9.Debezium Custom Converter](#quarkus-debezium-custom-converter)
+
+### Quarkus Debezium Engine Instance
+
+The proposal introduces an interface `Debezium` as an abstraction that exposes the engine's snapshot status.
+
+```java
+/**
+ * The Debezium engine abstraction in the Quarkus CDI
+ * <p>
+ * The Engine is submitted to an {@link Executor} or {@link ExecutorService} for execution by a single thread
+ */
+public interface Debezium {
+
+    /**
+     * @return engine's signaler, if it supports signaling
+     * @throws UnsupportedOperationException if signaling is not supported by this engine
+     */
+    Signaler signaler();
```

Comment:
@kmos I wonder if we could expose Signaller also via CDI events. So instead of direct access to signaller a signal event would be emitted.

Reply:
Something like:

```java
import io.debezium.engine.InsertEvent;
import jakarta.enterprise.context.ApplicationScoped;

@ApplicationScoped  
class Signaler {  
    
    public void handler(@Observes Signal signal) {
        /// some logic to apply
    }
    
}
```

@jpechane ?

Reply:
ok so something like:

```java
import io.debezium.engine.InsertEvent;
import jakarta.enterprise.context.ApplicationScoped;

@ApplicationScoped  
class SomethingThatWantToSignal {
   @Inject
   Signaler signaler;
    
    public void produce() {
        /// some logic
       signaler.signal(new DebeziumEngine.Signal("1", "log", "{\"message\": \"Signal message at offset ''{}''\"}", null));
    }
    
}
```

Reply:
something in the *quarkus way* to produce event:

```java
import io.debezium.engine.InsertEvent;
import jakarta.enterprise.context.ApplicationScoped;

@ApplicationScoped  
class SomethingThatWantToSignal {
   @Inject
   Event<Signal> event;
    
    public void produce() {
        /// some logic
       event.fire(new DebeziumEngine.Signal("1", "log", "{\"message\": \"Signal message at offset ''{}''\"}", null));
    }
    
}
```